MoreLinq.Test/MemoizeTest.cs

```diff
@@ -0,0 +1,193 @@
+            {
+                flow.Add(i.ToString());
+
+                if (i == 3 || i == 7)
+                {
+                    //consume 1-3 already cached
+                    //add 4-5 to cache (so go to outer loop)
+                    //consume 1-7 already cached
+                    //add 8-10 to cache (so go to outer loop)
+
+                    flow.Add("enter inner loop");
+                    foreach (var j in source)
+                    {
+                        flow.Add(j.ToString());
+
+                        if (!firstVisitAtInnerLoopDone && j == 5)
+                        {
+                            firstVisitAtInnerLoopDone = true;
+                            break;
+                        }
+                    }
+                    flow.Add("exit inner loop");
+                }
+            }
+            flow.Add("exit outer loop");
+
+            flow.Add("enter last loop");
+            //consume 1-10 (all item were already cached)
+            foreach (var k in source)
+            {
+                flow.Add(k.ToString());
+            }
+            flow.Add("exit last loop");
+
+            return flow;
+        }
+
+        [Test]
+        public void MemoizeWithPartialIterationBeforeCompleteIteration()
+        {
+            var buffer = Enumerable.Range(0, 10).Memoize();
+
+            buffer.Take(5).AssertSequenceEqual(Enumerable.Range(0, 5));
+            buffer.AssertSequenceEqual(Enumerable.Range(0, 10));
+        }
+
+        [Test]
+        public void MemoizeIsLazy()
+        {
+            new BreakingSequence<int>().Memoize();
+        }
+
+        [Test]
+        public void MemoizeWithInMemoryCollection()
+        {
+            var list = Enumerable.Range(1, 10).ToList();
+
+            Assert.IsInstanceOf<List<int>>(list.Memoize());
+        }
+
+        [Test]
+        public void MemoizeEnumeratesOnlyOnce()
+        {
+            var memoized = Enumerable.Range(1, 10).AsTestingSequence().Memoize();
+
+            Assert.IsTrue(memoized.ToList().Count == 10);
+            Assert.IsTrue(memoized.ToList().Count == 10);
+        }
+
+        [Test]
+        public void MemoizeDoesNotDisponseOnEarlyExitByDefault()
+        {
+            Assert.Throws<AssertionException>(() =>
+            {
+                using (var xs = new[] { 1, 2 }.AsTestingSequence())
+                {
+                    xs.Memoize().Take(1).Consume();
+                    xs.Memoize(false).Take(1).Consume();
+                }
+            });
+        }
+
+        [Test]
+        public void MemoizeWithDisponseOnEarlyExitTrue()
+        {
+            using (var xs = new[] { 1, 2 }.AsTestingSequence())
+                xs.Memoize(true).Take(1).Consume();
+        }
+
+        [Test]
+        public void MemoizeDisponsesAfterSourceIsIteratedEntirely()
+        {
+            using (var xs = new[] { 1, 2 }.AsTestingSequence())
+                xs.Memoize().Consume();
+        }
+
+        [Test]
```

Comment:
Those numbers are on a Windows 10 box. On a mac OS, the tests running under Mono get far worse, 5.1 vs. 46.1 seconds.